Feature/relative matching distance

chahank opened the pull request:
Changes proposed in this PR:
- This PR contains the changes from #1020 
- In addition, it adds a relative matching threshold based on the equivalent grid spacing of the centroids.

This PR fixes #1079.

### PR Author Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] Correct target branch selected (if unsure, select `develop`)
- [ ] Descriptive pull request title added
- [ ] Source branch up-to-date with target branch
- [ ] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [ ] [Tests][testing] updated
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

peanutfun commented on climada/util/coordinates.py:
```diff
@@ -85,6 +85,79 @@
+    test : bool
+        True if lat/lon ranges seem to be in the geographic coordinates range, otherwise False.
+    """
+    lat = np.array(lat)
+    lon = np.array(lon)
```
This creates copies. Use `asarray` to avoid copies if numpy arrays are passed

peanutfun commented on climada/util/coordinates.py:
```diff
@@ -85,6 +85,79 @@
+
+    Returns
+    -------
+    test : bool
+        True if lat/lon ranges seem to be in the geographic coordinates range, otherwise False.
```
This function never returns `False`. It either returns `True` or raises an error. I think you should split that functionality: Have a function `is_geo_coords` (or similar) that returns true/false, and then have a function `check_XZY`, that raises an error or returns None.

chahank replied:
Done.

peanutfun commented on climada/util/coordinates.py:
```diff
@@ -85,6 +85,79 @@
+            " If you use degree values outside of these ranges,"
+            " please shift the coordinates to the valid ranges."
+        )
+    return True
```
?
```suggestion
    return like_geo
```

peanutfun commented on climada/util/coordinates.py:
```diff
@@ -1059,10 +1147,13 @@ def match_coordinates(
         assigned_idx = np.arange(coords.shape[0])
     else:
+        if threshold is None:
+            threshold = 2 * max(abs(r) for r in get_resolution(coords_to_assign.T))
```
`get_1d_resolution` can be extremely costly. It takes unique coordinates and computes differences between them. Shouldn't we raise a warning that this is suboptimal?

https://github.com/CLIMADA-project/climada_python/blob/08ec51bb958436cd18e77cf8f4178bc05a295d78/climada/util/coordinates.py#L1836

chahank replied:
Testing, for a regular grid with 65'000 points (world at 1° resolution), it takes 13ms, for 6'500'000 points (world at 0.1° resolution) it takes 1.5s. Is this acceptable?

<img width="603" height="411" alt="Screenshot 2025-08-27 at 15 07 09" src="https://github.com/user-attachments/assets/9be941a9-2aa2-4065-93ef-7731c4f1767b" />

peanutfun replied:
Yes, that seems acceptable to me! Thanks a lot for the testing 🙌

peanutfun commented on climada/util/test/test_coordinates.py:
```diff
@@ -230,6 +236,47 @@ def def_ref_50():
 
 
+class TestGetUnitCoords(unittest.TestCase):
+    def create_mock_gdf(self, crs, num_points=5):
+        geometry = [
+            shapely.Point(x, y) for x, y in zip(range(num_points), range(num_points))
+        ]
+        gdf = gpd.GeoDataFrame(geometry=geometry, crs=crs)
+        return gdf
```
This is called a *test fixture*. Please use `setUp` to create fixtures and save them in the `TestCase` object. You can then retrieve them from the `self` object within each test case. See https://docs.python.org/3/library/unittest.html#organizing-test-code

peanutfun replied:
Indeed, I was wrong about the usage of a fixture here. What you do is more of a [parameterized test](https://docs.pytest.org/en/7.1.x/how-to/parametrize.html#parametrize-basics), which unittest cannot really facilitate 🤷

peanutfun reviewed (changes requested):
Stumbled upon this and had the time, so I left a review. This looks good already, but I found some issues with data being copied and default operations that might be very costly.

chahank commented:
Ok, all tests are passing, I think the implementation works now. 

TO DO:
- work-in the comments
- improve / adjust all the docstrings where needed
- update the documentation

chahank commented:
@luseverin : it would be good that you review my changes once to check functionality as you made the original PR https://github.com/CLIMADA-project/climada_python/pull/1020 this is based upon.

chahank commented:
Note: the method `lon_normalize` is apparently used in several tests to change longitudinal values that do not conform to the constraints of the new `check_if_geo_coordinates` method. Hence, one can either accept that `lon_normalize` does also work with non-conform coordinates, or one would have to update the several tests in python core and petals. @luseverin ?

luseverin reviewed (commented):
Thanks @chahank! I only have some minor comments: 

- First a clarification; I am not exactly sure of what is supposed to happen in the case of non-regular coordinates for the determination of the distance threshold, and I am also a bit confused with the "highest resolution" formulation in the description of the threshold definition, see my comments.
- Some tests are missing in  `assign_centroids` (when crs are not equal), in `get_gridcellarea` (when the unit is unknown), in `get_crs_unit`, in `dist_to_coast`, and in `estimate_matching_threshold`.

Otherwise, based on your changes and the tests, it seems to be in order in terms of functionality.

chahank commented:
> > Note: the method `lon_normalize` is apparently used in several tests to change longitudinal values that do not conform to the constraints of the new `check_if_geo_coordinates` method. Hence, one can either accept that `lon_normalize` does also work with non-conform coordinates, or one would have to update the several tests in python core and petals. @luseverin ?
> 
> Is that an in issue? As far as I understand, the `check_geo_coordinates` function tests that the input coordinates are likely geographic, whereas the `lon_normalize` only makes sure that input coordinates (in degrees) are normalized to a -180<lon<180 range. So in the end, they do different things, but it is still consistent as we still want that the input coordinates are likely geographic before attempting any normalization. Or did I misunderstand that? Maybe you could indicate in which tests this is an issue.

`lon_normalize` idea is that if coordinates are say 0-360, they are moved to -180, 180.  However, the method now accepts any values and simply normalizes them to be in the -180 to 180 range. Thus faulty meter values can be entered, or also faulty degree ranges. The latter is done in several tests, and has also led to incorrect results in the past. Basically, you can have values say 0 to 400. Do we want to prevent this?

chahank commented:
> Ok, I see. Then it means that we should be raising a `ValueError` with the `check_if_geo_coords` function in those instances. Should we simply call `check_if_geo_coords` within `lon_normalize` and adapt the tests accordingly?

I think it is really worth trying. I tried it quickly, and many tests failed, and I did not yet formed an opinion whether this means that we should adapt the tests or do something else.

chahank commented:
> Ok so I added `check_if_geo_coords` in `lon_normalize` and updated the couple tests that could be easily updated (essentially, removed test data that was not consistent with our definition of geographic coordinates). A few other things happened in addition:
> 
>     * `test_set_one_file_pass` was failing because an empty lon array was inputed to `lon_normalize`within the call to `tropCyclone.from_tracks(tc_track, centroids=CENTR_TEST_BRB) `. I fixed it by adding an explicit handling of empty arrays in `is_geo_coords`, where empty arrays are considered as valid geographic coordinates. 

Excellent!

> I do not know if this is a desirable feature, and I also do not know if the empty lon array in `tropCyclone.from_tracks(tc_track, centroids=CENTR_TEST_BRB) ` is a bug or not. It seems that the `idx_centr_filter` variable becomes an empty array at some point, leading to an empty centroids array, but I did not investigate this further..
> 
>     * In `test_ibtracs_with_basin`, the track downloaded with` tc_track = tc.TCTracks.from_ibtracs_netcdf(year_range=(2002, 2003), basin="WP", genesis_basin="EP")` has longitudinal value that have extent > 360 (lon_min=-179, lon_max=182.6). I increased the tolerance in `is_geo_coords` for the longitudinal extent from 361 to 541 as a fix. It seems to be working for now and I hope this should work for all tracks that one can get from IBtracks but this is only a guess..

Great. Weird that IBTrACS basically has invalid coordinates, but it works ;). 
Could once try to load all IBTrACS?

> 
>     * A couple tests from the LitPop module are now failing:  `test_switzerland300_pass` and `test_Liechtenstein_15_lit_pass`. It seems that some parameters of the Affine rasterio transform have been changed, e.g. `ent.meta["transform"][4] == 0.08333333000000209` instead of `-0.08333333333333333`, and `ent.meta["transform"][5] == 45.833333335` instead of `47.75`. @chahank any idea where this might come from?

Was this changed in this PR? Then it might have been a mistake on my side. Please feel free to revert these changes.

chahank commented:
Oh, one last thing we should do, since this affects the core climada computation, is to check whether the tutorials still work.
